macros/src/types/named.rs

```diff
@@ -102,27 +102,35 @@ fn format_field(
         return Ok(());
     }
 
-    let parsed_ty = field_attr.type_as(&field.ty);
-
-    let (ty, optional_annotation) = match field_attr.optional {
-        Optional {
-            optional: true,
-            nullable,
-        } => {
-            let inner_type = extract_option_argument(&parsed_ty)?; // inner type of the optional
-            match nullable {
-                true => (&parsed_ty, "?"),  // if it's nullable, we keep the original type
-                false => (inner_type, "?"), // if not, we use the Option's inner type
+    let ty = field_attr.type_as(&field.ty);
+
+    let opt = match (struct_optional, field_attr.optional) {
+        (opt @ Optional::Optional { .. }, Optional::NotOptional) => opt,
+        (_, opt @ Optional::Optional { .. }) => opt,
+        (opt @ Optional::NotOptional, Optional::NotOptional) => opt,
+    };
+
+    let optional_annotation = if let Optional::Optional { .. } = opt {
+        quote! { if <#ty as #crate_rename::TS>::IS_OPTION { "?" } else { "" } }
+    } else {
+        quote! { "" }
+    };
+
+    let optional_annotation = if let Optional::Optional { .. } = field_attr.optional {
+        quote! {
+            if <#ty as #crate_rename::TS>::IS_OPTION {
+                #optional_annotation
+            } else {
+                panic!("`#[ts(optional)]` can only be used with the Option type")
             }
         }
-        Optional {
-            optional: false, ..
-        } => (&parsed_ty, ""),
+    } else {
+        optional_annotation

```

Comment:
Another option is just remove this whole thing, this way there'll be no panic (still, no compiler error). What will happen is that `#[ts(optional)]` will just do nothing when not using an Option type